architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,329 @@
+- 
+### Goals
+
+- Support users defining their own resource types and resource type schemas for use with Radius.
+	- Support rich declarative validation of resource request bodies based on schema.
+	- Support versioning for user-defined resource types.
+		- Including the ability to support multiple API versions.
+		- Including the ability to introduce breaking changes between versions.
+	- eg: Users can define a new resource type and it's schema. Radius will use the schema for validation. Users can add new versions of the resource type, and old ones continue to function.
+- Support dynamic management of resource types.
+	- eg: Users can register and unregister resource types.
+	- eg: Tooling like the Dashboard and `rad` CLI dynamically query the set of resource types rather than hardcoding a list of known types.
+- Support use of user-defined types in Bicep with full fidelity.
+	- eg: Users can define their own types and use them in their Bicep application definitions.
+- Support use of user-defined types as part of an application with full fidelity.
+	- User defined types can opt-in to support connections, recipes, application-graph.
+	- eg: Users can define their own types and incorporate them in applications. They work the same as existing types that weren't defined by the user.
+- Support multiple implementation choices for user-defined types.
+	- User-defined types should be easy to implement using declarative functionality.
+	- User-defined types should be powerful enough to support implementing a resource-provider.
+- Cultivate an open-source ecosystem for user-defined types.
+	- Users can find documentation for how to define and publish user-defined types either through open-source or internally to their organization.
+	- Users can find prescriptive guidance to simplify the design of resource schemas. 
+	- Users have a productive inner-loop for developing resource types.
+	- The Radius project maintains a community types repository with infrastructure for testing, publishing, and contributing.
+	- The Radius project defines a maturity model for a type graduating from the *community* repository to official support as part of the `Applications.[xyz]` namespace.
+- The Radius project heavily leverages the user-defined types infrastructure for the implementation of Radius.
+	- All system-defined types provide a schema through the same registration mechanism as user-defined types. 
+	- Most of the currently system-defined types migrate to use user-defined types. We will allow exceptions for resource types that have unique requirements like `Applications.Core/secretStores`.
+
+### Non goals
+
+- (non-goal): Migrate **all** existing resource types. 
+	- Some types in Radius have special behaviors that are intentionally unique.
+	- eg: `Applications.Core/secretStores` has a unique set of capabilities for interacting with secrets. We should probably not expand the scope of user-defined types to include this because it's complex and not valuable.
```

Comment:
Could you give some examples of the capabilities that make it complex?

Reply:
The part where it stores secrets in an external secret store (Kubernetes for now) is complex. This requires imperative code on the request-path, which isn't part of the UDT design described here.